Add library panel repeats to the grid item during V2alpha1 -> V2beta1 conversion.

oscarkilhed opened the pull request:
This is an attempt to add the repeats from library panels to the grid items of the new schema version.

When converting we will go through the grid items, look up the element in the elements collection and use the library panel ref to look up the library panel using the library panel service. We then assign the repeat options from the library panel to the grid item.

Currently the problem is we're assuming OrgID = 1. I'm not sure how to get the correct OrgID in this context, and without it we can't look up the library panel.

ivanortegaalba committed:
SchemaV2: Convertion from v1beta1 to v2beta1

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

ivanortegaalba committed:
Compare backend-frontend v1 convertion

ivanortegaalba committed:
Compare backend-frontend v1 convertion

ivanortegaalba committed:
Fix fe be diff

ivanortegaalba committed:
Resolve DS issues

ivanortegaalba committed:
Fix ds inconsistecnies

harisrozajac committed:
Merge main

harisrozajac committed:
fix legacy string value issues

harisrozajac committed:
fix ds test

harisrozajac committed:
fix layout issue

harisrozajac committed:
update test

oscarkilhed committed:
wip: initial attempt

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

# Conflicts:
#	apps/dashboard/pkg/migration/conversion/conversion.go
#	apps/dashboard/pkg/migration/conversion/v1.go

ivanortegaalba committed:
Fix tests and issue with defaultConfig

ivanortegaalba committed:
Update output

ivanortegaalba committed:
Fix viz config convertion

ivanortegaalba committed:
wip

ivanortegaalba committed:
Merge origin/main into iortega/transform-v1-to-v2

Resolved conflicts in:
- apps/dashboard/pkg/migration/conversion/conversion.go: Updated conversion functions to use withConversionMetrics
- pkg/registry/apis/dashboard/register.go: Updated migration initialization
- pkg/storage/unified/testing/storage_backend.go: Fixed test helper function
- public/app/features/dashboard/state/DashboardMigratorToBackend.test.ts: Updated variable adapters registration
- Removed .betterer.results file as it was deleted in origin/main

ivanortegaalba committed:
Fix v1 to v2 dashboard transformation differences

Major fixes implemented:
- Backend function names in conversion.go
- Backend group field logic for queries, annotations, and vizConfig
- Backend datasource resolution with map-based lookup
- Backend timezone handling (empty string vs browser)
- Backend annotation processing (empty array vs default annotation)
- Backend default values (editable, liveNow)
- Backend variable processing (definition, defaultKeys, refresh, refId)
- Backend panel layout (y position calculations)
- Backend VizConfig (Kind and Group fields, default values)
- Frontend snapshot issue (annotations not processing)
- Frontend datasource references (only when original has valid datasource)

Test results:
- annotation-conversions: PASSING (0 differences)
- dashboard-properties: 3 expected architectural differences
- panel-conversions: Multiple expected architectural differences
- variable-conversions: 7 expected architectural differences

All remaining differences are expected architectural choices between
backend persistence optimization and frontend UI consumption optimization.

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
fix issues with panel and annotation queries with no datasource

harisrozajac committed:
definition and regex

harisrozajac committed:
Use proper v1beta1 resource when testing

harisrozajac committed:
remove misc file

harisrozajac committed:
fix ds provider test

harisrozajac committed:
fix def ds test in response transformer

harisrozajac committed:
fix remaining ResponseTransformers test

harisrozajac committed:
timesettings, variable refresh, editable, liveNow, definition

harisrozajac committed:
fix transformSceneToSaveModelSchemaV2 test

harisrozajac committed:
revert legacyRow changes

harisrozajac committed:
fix go lint issues

harisrozajac committed:
normalize y coordinates when serializing a row

harisrozajac committed:
clean up

harisrozajac committed:
update tests

harisrozajac committed:
use GetStringValue from schemaversion

harisrozajac committed:
fix go lint - cyclomatic complexity

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
update open api snapshot

harisrozajac committed:
add migrated dashboards

harisrozajac committed:
fix default panel type when panel type is not provided

harisrozajac committed:
revert dash link changes for now

harisrozajac committed:
fix

harisrozajac committed:
fix nested panel issue and default ref in v1

harisrozajac committed:
apply defaults to nested panels too

harisrozajac committed:
update snapshots

harisrozajac committed:
fix issues with annotations

harisrozajac committed:
matchers, showLegend, annotations

harisrozajac committed:
when converting also don't process queries that have only a refId

harisrozajac committed:
fix issues with text var

harisrozajac committed:
fix dash links

harisrozajac committed:
default to collapse: false when serializing

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

ivanortegaalba committed:
fix: filter refId from variable query specs in backend migration

- Add buildDataQueryKindForVariable function to filter refId for variables
- Remove default refId "A" in transformSingleQuery
- Only include __legacyStringValue for non-empty string queries
- Remove refId addition in transformSaveModelSchemaV2ToScene.getDataQueryForVariable
- Handle undefined queries gracefully in frontend and backend
- Ensure backend matches frontend behavior for query variable serialization

ivanortegaalba committed:
fix: default variable refresh to 'never' to match frontend behavior

Change backend default for missing refresh field from 'onDashboardLoad'
to 'never' to match frontend defaultVariableRefresh() schema default

ivanortegaalba committed:
fix: only include iconColor in annotations when it exists

- Frontend: Use defaultAnnotationQuerySpec().iconColor as fallback to match schema defaults
- Backend: Only set iconColor if it exists in v1 input (not using GetStringValue)
- Ensures iconColor is only included when present in original dashboard

ivanortegaalba committed:
fix: use schema defaults for annotation enable, hide, and iconColor

- Use defaultAnnotationQuerySpec() to get schema defaults instead of hardcoded values
- Default enable to false (schema default) to match frontend behavior
- Use schema default for iconColor and hide fields
- Ensures consistency with frontend which uses defaultAnnotationQuerySpec() defaults

ivanortegaalba committed:
fix: set collapse for hidden-header rows to match first explicit row

- When panels appear before the first explicit row, the hidden-header row's
  collapse should match the first explicit row's collapsed value
- Matches frontend behavior where collapse: panel.collapsed uses the next
  row panel's collapsed value
- Ensures consistency between frontend and backend when converting rows layout

ivanortegaalba committed:
fix: handle constant variables with missing query value

- Frontend: Fix bug where undefined value was converted to string 'undefined'
  - Now defaults to empty string when value is undefined: value ? String(value) : ''
- Backend: Match frontend fix - default to empty string for text/value when query is missing
- Ensures consistency when constant variable query is missing from v1 dashboard

ivanortegaalba committed:
Fix interval variable handling when query is missing

- Extract intervals from options when query is missing/empty (matches backend behavior)
- Handle undefined/null query in getIntervalsFromQueryString
- Handle missing current object/value in getCurrentValueForOldIntervalModel
- Update interval variable refresh to use literal 'onTimeRangeChanged' in schema
- Use defaultIntervalVariableSpec() for interval variable serialization
- Backend: Generate query string from options when query is missing

ivanortegaalba committed:
Fix corrupted dashboard with systemRef override

harisrozajac committed:
don't resolve types for template variables in datasource refs on the backend

harisrozajac committed:
fix annotation and ds issues

harisrozajac committed:
fix range and special mappings

harisrozajac committed:
fix datasource var pluginId and regex

harisrozajac committed:
add __systemRef to schema

harisrozajac committed:
update v15 migration annotation to have a ds type because v2 keeps track of if type is in the initial save model, and if it's not it removes it, but for frontendOuput we are running transformSaveModelToScene which will then assign the type

harisrozajac committed:
add migration fields since the backend applies automigrations in collapsed rows

harisrozajac committed:
filter out queries in ResponseTransformer that only have refId field

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
lint

harisrozajac committed:
v2: add default query if queries are empty to match v1 behavior

harisrozajac committed:
fix single migration test

harisrozajac committed:
tracking test should have a defined spec otherwise datasource is removed and won't be tracked

harisrozajac committed:
initialize default with default ds ref

harisrozajac committed:
wip

ivanortegaalba committed:
Do not assign DS if ds group is empty

harisrozajac committed:
cleanup

harisrozajac committed:
revert change in setupTests.ts

harisrozajac committed:
clean up TODO

harisrozajac committed:
query with only refId should not expect to have a group

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

ivanortegaalba committed:
refactor: extract v0alpha1 to v1beta1 conversion logic into atomic function

- Extract ConvertDashboard_V0_to_V1beta1 into v0alpha1_to_v1beta1.go
- Extract prepareV0ConversionContext and migrateV0Dashboard helper functions
- Standardize v0.go to match v1.go pattern with inline multi-step conversions
- Implement Convert_V0_to_V2alpha1 using atomic functions (v0->v1beta1->v2alpha1)
- Implement Convert_V0_to_V2beta1 using atomic functions (v0->v1beta1->v2alpha1->v2beta1)
- Remove non-atomic v0alpha1_to_v2alpha1.go file

ivanortegaalba committed:
test: add version-specific test files for conversion error handling

- Extract v0 conversion tests into v0_test.go
- Extract v1 conversion tests into v1_test.go
- Add v2 conversion tests in v2_test.go
- Ensure all error handling paths in conversion functions are covered
- Add tests for Convert_V0_to_V2alpha1 and Convert_V0_to_V2beta1 error paths
- Add tests for Convert_V1beta1_to_V2alpha1 and Convert_V1beta1_to_V2beta1 error paths
- Add tests for Convert_V2alpha1_to_V2beta1 error handling

ivanortegaalba committed:
Fix tests

ivanortegaalba committed:
Fix linter

ivanortegaalba committed:
Clean up

harisrozajac committed:
feat(dashboard): Add automatic data loss detection for dashboard conversions

Implements comprehensive data loss detection for all dashboard API version conversions.

Components Tracked:
• Panels (visualization + library panels)
• Queries (data source queries, excludes row panel queries)
• Annotations
• Links
• Variables (template variables)

Features:
• Automatic detection via withConversionMetrics wrapper (zero code changes)
• Error type: 'conversion_data_loss_error'
• Logs: panelsLost, queriesLost, annotationsLost, linksLost, variablesLost

Bugs Found:
• Fixed critical bug: metrics.go was silently swallowing ALL errors (return nil → return err)

Testing:
• TestDataLossDetectionOnAllInputFiles - runs all conversions with detailed logging
• V2→V0/V1 downgrades write output for debugging then skip (not yet implemented)
• All tests passing

harisrozajac committed:
update data loss tests to use target and source versions

harisrozajac committed:
go lint

harisrozajac committed:
cleanup

harisrozajac committed:
Remove unused field

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
withConversionMetrics need to return nil unconditionally to avoid 500 server error

harisrozajac commented on apps/dashboard/pkg/migration/testutil/fakes.go:
Let's move this under `apps/dashboard/pkg/migration/testutil/mocks.go`

harisrozajac commented on apps/dashboard/pkg/migration/conversion/v2alpha1_to_v2beta1.go:
```diff
@@ -34,9 +37,37 @@ import (
+		GetPanelModelByUID(c context.Context, signedInUser identity.Requester, uid string) (map[string]interface{}, error)
+	}
+	if mg, ok := any(svc).(modelGetter); ok {
+		ctx, ident := identity.WithServiceIdentity(context.Background(), 1) // Default orgID to 1, not sure how to get the correct orgID when doing conversions.
```
Since you are already using identity service here and that service has GetOrgID method, something like this should work to get orgID:
```go
user, err := identity.GetRequester(ctx)
if err != nil {
     return fmt.Errorf("error getting requester: %w", err)
    }
orgId := user.GetOrgID()
```

harisrozajac reviewed (commented):
Looks good overall. I left some comments. We'll need to resolve merge conflicts before adding it to feature branch.

harisrozajac committed:
Update public/app/features/dashboard-scene/utils/utils.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>

harisrozajac committed:
Update public/app/features/dashboard-scene/serialization/layoutSerializers/utils.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
lint

oscarkilhed committed:
fix v2alpha1 conversion overwriting layout

harisrozajac committed:
regenerate snapshots

ivanortegaalba committed:
Apply suggestions from code review

Co-authored-by: Stephanie Hingtgen <stephanie.hingtgen@grafana.com>

ivanortegaalba committed:
Remove unnecessary docs

ivanortegaalba committed:
Update apps/dashboard/pkg/migration/README.md

Co-authored-by: Stephanie Hingtgen <stephanie.hingtgen@grafana.com>

ivanortegaalba committed:
Add comment

ivanortegaalba committed:
Reuse migrator dsprovider and put it in ctx

ivanortegaalba committed:
Fix issue with pluginId using UID instead of type

oscarkilhed committed:
feat: attach library panel repeat options during v2alpha1 to v2beta1 conversion

- Add library panel service support to conversion package
- Create librarypanel.go for library panel service context handling
- Update v2alpha1_to_v2beta1.go to attach repeat options from library panels
  to grid and auto-grid layout items during conversion
- Simplify prepareV2alpha1ConversionContext to match v1beta1 pattern
- Update register.go to initialize library panel service for conversions
- Regenerate test output files for library panel repeat tests

This ensures that library panel repeat options (repeat variable, direction,
and maxPerRow) are preserved when converting dashboards from v2alpha1 to
v2beta1 schema.

oscarkilhed committed:
refactor: remove duplicate FakeLibraryPanelService

Remove duplicate FakeLibraryPanelService from migration/testutil/fakes.go.
The conversion tests use the one from conversion/testutil/mocks.go, so
this duplicate was unused and causing confusion.

oscarkilhed committed:
Remove ds info provider from conversion context
